baseband/guppi/frame.py

```diff
@@ -0,0 +1,201 @@
+
+    A number of properties are defined: `shape`, `dtype` and `size` are
+    the shape, type and number of complete samples of the data array, and
+    `nbytes` the frame size in bytes.  Furthermore, the frame acts as a
+    dictionary, with keys those of the header.  Any attribute that is not
+    defined on the frame itself, such as ``.time`` will be looked up on the
+    header as well.
+    """
+    _header_class = GUPPIHeader
+    _payload_class = GUPPIPayload
+
+    @classmethod
+    def fromfile(cls, fh, memmap=True, valid=True, verify=True):
+        """Read a frame from a filehandle, possible mapping the payload.
+
+        Parameters
+        ----------
+        fh : filehandle
+            To read header from.
+        memmap : bool, optional
+            If `True` (default), use `~numpy.memmap` to map the payload.
+            If `False`, just read it from disk.
+        valid : bool, optional
+            Whether the data are valid (default: `True`). Note that this cannot
+            be inferred from the header or payload itself.  If `False`, any
+            data read will be set to ``cls.fill_value``.
+        verify : bool, optional
+            Whether to do basic verification of integrity.  Default: `True`.
+        """
+        header = cls._header_class.fromfile(fh, verify=verify)
+        payload = cls._payload_class.fromfile(fh, header=header, memmap=memmap)
+        return cls(header, payload, valid=valid, verify=verify)
+
+    @classmethod
+    def fromdata(cls, data, header=None, valid=True, verify=True, **kwargs):
+        """Construct frame from data and header.
+
+        Note that since GUPPI files are generally very large, one would
+        normally map the file, and then set pieces of it by assigning to slices
+        of the frame.  See `~baseband.guppi.base.GUPPIFileWriter.memmap_frame`.
+
+        Parameters
+        ----------
+        data : `~numpy.ndarray`
+            Array holding complex or real data to be encoded.
+        header : `~baseband.dada.GUPPIHeader` or None, optional
+            If not given, will attempt to generate one using the keywords.
+        valid : bool, optional
+            Whether the data are valid (default: `True`). Note that this
+            information cannot be written to disk.
+        verify : bool, optional
+            Whether or not to do basic assertions that check the integrity.
+            Default: `True`.
+        **kwargs
+            If ``header`` is not given, these are used to initialize one.
+        """
+        if header is None:
+            header = cls._header_class.fromvalues(verify=verify, **kwargs)
+        payload = cls._payload_class.fromdata(data, header=header)
+        return cls(header, payload, valid=valid, verify=verify)
+
+    def __len__(self):
+        """Number of samples, subtracting those in the overlap region."""
+        return self.header.samples_per_frame
+
+    def _get_payload_item(self, item):
```

Comment:
Do we actually need this? The stream readers will always pass on an actual start and stop, which will already be limited by the frame size. I'm wondering whether the choice of doing the cutting here is in the end the best idea. Overall, I feel the frame doesn't really have to care about overlap, that is up to the stream.

Reply:
Agreed that we don't need this.  I've now moved handling the overlap to stream readers.